### DISCUSSION TITLE
07_paraview  /Docker/ Example Issues

### DISCUSSION THREAD

**Discussion by PierreLouis94** [Kitware/trame#578] [Trame/ParaView] [answered]
After building this dockerfile

      FROM kitware/trame:py3.10-ubuntu22.04
      
      RUN install -d -o trame-user -g trame-user /deploy
      RUN apt update -y && apt install -y libosmesa6-dev
      
      #Only changed the paraview link because the one in the example is not working
      ARG PV_URL='https://www.paraview.org/files/v5.12/ParaView-5.12.0-osmesa-MPI-Linux-Python3.10-x86_64.tar.gz'
      RUN mkdir -p /opt/paraview && cd /opt/paraview && wget -qO- $PV_URL | tar --strip-components=1 -xzv
      ENV TRAME_PARAVIEW=/opt/paraview
      
      COPY --chown=trame-user:trame-user . /deploy
      
      RUN /opt/trame/entrypoint.sh build

and running it with this `docker run -it --rm -p 8080:80 my-trame-app`

i get 

    (paraview3.10) [$][~/trame/examples/07_paraview/Docker$ docker run -it --rm -p 8080:80 my-trame-app
     * Restarting Apache httpd web server apache2                                                                                                                   AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 172.17.0.2. Set the 'ServerName' directive globally to suppress this message
                                                                                                                                                             [ OK ]
    Starting server...
    Starting the wslink launcher at
    ======== Running on http://0.0.0.0:9000 ========
    (Press CTRL+C to quit)

at first sight it seems to run, but all i see is this.

![Screenshot from 2024-08-29 00-44-59](https://github.com/user-attachments/assets/821b9e64-fdee-4125-be7d-7826451d6a5f)

and it's not changing. If i try to reload the page i get these error logs.

      Error handling request
      Traceback (most recent call last):
        File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 462, in _handle_request
          resp = await request_handler(request)
        File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_app.py", line 537, in _handle
          resp = await handler(request)
        File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 97, in handle_post
          return await self._waitForReady(session, request)
        File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 109, in _waitForReady
          if self.process_manager.isReady(session, count):
        File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 480, in isReady
          if not self.isRunning(id) and count < 60:
        File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 468, in isRunning
          return self.processes[id].poll() is None
      KeyError: '4e2ae9d0-658f-11ef-8ae6-0242ac110002'

I managed to containerize multiple files trame applications with vue-components thanks to cookiecutter. But I failed to containerize anything yet where paraview is involved. My end goal is to containerize the visualizer but the dockerfile, the scripts and the base images all seem outdated. Could anyone help my find the issue here with this simple paraview docker example?

**Comment by jourdain** [comment]
Try that 

```
FROM kitware/trame:py3.10-glvnd
  
RUN install -d -o trame-user -g trame-user /deploy
  
ARG PV_URL='https://www.paraview.org/files/v5.13/ParaView-5.13.0-egl-MPI-Linux-Python3.10-x86_64.tar.gz'
RUN mkdir -p /opt/paraview && cd /opt/paraview && wget -qO- $PV_URL | tar --strip-components=1 -xzv
ENV TRAME_PARAVIEW=/opt/paraview
  
COPY --chown=trame-user:trame-user . /deploy
  
RUN /opt/trame/entrypoint.sh build
```

This should give you the option to run with or without gpu

**Reply by PierreLouis94** [reply]
Updating my Dockerfile to this still isn't working. I'm stuck on the loading screen. The only difference now is that when I reload, I no longer get the error messages from the previous Dockerfile.

**Reply by jourdain** [reply]
Can you provide the content of the session log? Inside `/deploy/server/logs/launcher/{xxxxxxxxxxxxx}.txt`

**Reply by PierreLouis94** [reply]
```
root@f6e21b105531:/deploy/server/logs/launcher# ls
6710800e-6963-11ef-896a-0242ac110002.txt  67e0b1d4-6963-11ef-896a-0242ac110002.txt  launcherLog.log
root@f6e21b105531:/deploy/server/logs/launcher# cat 6710800e-6963-11ef-896a-0242ac110002.txt
ParaView is using venv: /deploy/server/venv

App running at:
 - Local:   http://0.0.0.0:9500/
 - Network: http://0.0.0.0:9500/

Note that for multi-users you need to use and configure a launcher.
root@f6e21b105531:/deploy/server/logs/launcher# cat 67e0b1d4-6963-11ef-896a-0242ac110002.txt
ParaView is using venv: /deploy/server/venv

App running at:
 - Local:   http://0.0.0.0:9499/
 - Network: http://0.0.0.0:9499/

Note that for multi-users you need to use and configure a launcher.
```
These are the 2 txt files that were inside the path you mentioned. Seems like paraview is running on the venv like it should right?

**Reply by jourdain** [reply]
Any error on the JS side? Is it because of the wrong client type? The default now is vue3. If your client is vue2, you need to add that to the docker definition.

```
ENV TRAME_CLIENT_TYPE=vue2
```

**Reply by PierreLouis94** [reply]
Thank you very much. With your help, I was able to dockerize my first single-file Trame application with ParaView. Now I want to use the Dockerfile to make a cookiecutter app with components that work with ParaView. I added `from paraview import simple` into the core.py file to see if it would work with ParaView. It worked without it, but unfortunately, it didn’t work with the import of ParaView. The problem lies in the apps.yaml file. I'm not sure how to correctly configure it so that it runs with ParaView.

This is the standrad apps.yml file where I got the errors. 

```
trame: # Default app under /index.html
  www_modules:
    - trame_app.module
  app: trame-app
trame-app: # /trame-app.html
  app: trame-app
```
So with some inspiration of the visualizer I changed it to this:
```
trame: # Default app under /index.html
  www_modules:
    - trame_app.module
  cmd:
    - /opt/paraview/bin/pvpython
    - -m
    - paraview.apps.trame
    - --trame-app
    - trame-app
```
with this apps.yaml configuration apparently Paraview is using the right venv but somehow my app is not being recognized.

```
(base) [$][~/Desktop$ docker ps
CONTAINER ID   IMAGE       COMMAND                  CREATED          STATUS          PORTS                                     NAMES
752cae6101f7   trame-app   "/opt/trame/entrypoi…"   57 seconds ago   Up 56 seconds   0.0.0.0:8080->80/tcp, [::]:8080->80/tcp   distracted_lederberg
(base) [$][~/Desktop$ docker exec -it 7 /bin/bash
root@752cae6101f7:/# cd deploy/server/logs/launcher
root@752cae6101f7:/deploy/server/logs/launcher# ls
457b1d48-6a3d-11ef-9c76-0242ac110002.txt  48915308-6a3d-11ef-9c76-0242ac110002.txt  launcherLog.log
root@752cae6101f7:/deploy/server/logs/launcher# cat 457b1d48-6a3d-11ef-9c76-0242ac110002.txt
ParaView is using venv: /deploy/server/venv
Traceback (most recent call last):
  File "/opt/paraview/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/paraview/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/opt/paraview/lib/python3.10/site-packages/paraview/apps/trame.py", line 28, in <module>
    main()
  File "/opt/paraview/lib/python3.10/site-packages/paraview/apps/trame.py", line 23, in main
    module = importlib.import_module(TRAME_APP)
  File "/opt/paraview/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1004, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'trame-app'
```
I dont really understand the apps.yaml right now. What exactly am I doing wrong?

**Reply by jourdain** [reply, marked as answer]
`trame-app` is a script shortcut. That encapsulate the python command line...  (defined inside your python package definition)

The `cmd` in apps.yaml expect exactly the command you would run locally. So you can try locally to run a command like that (outside of docker) to see if that works. 

Moreover, `python -m trame-app` would not work because `trame-app` is not a valid package name. But maybe `trame_app` might work. And the complex pvpython command expect the same thing as a `-m arg`. So once you figure out how you can run your app like `python -m xxx`, you should be able to adjust your pvpython command line.

**Reply by PierreLouis94** [reply]
Thank you very much Sebastien for your help. I managed to make the visualizer run on docker now!

**Comment by HarryHan990915** [comment]
Hi , I met this issue too, when building docker of trame-slicer application. What is the reason causing this. I used app.yml to run the cmd but still has this issue.

Traceback (most recent call last):
  File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 510, in _handle_request
    resp = await request_handler(request)
  File "/deploy/server/venv/lib/python3.10/site-packages/aiohttp/web_app.py", line 569, in _handle
    return await handler(request)
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 97, in handle_post
    return await self._waitForReady(session, request)
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/backends/aiohttp/launcher.py", line 109, in _waitForReady
    if self.process_manager.isReady(session, count):
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 480, in isReady
    if not self.isRunning(id) and count < 60:
  File "/deploy/server/venv/lib/python3.10/site-packages/wslink/launcher.py", line 468, in isRunning
    return self.processes[id].poll() is None
KeyError: 'f3594ffe-555e-11f0-95f0-d22cec25efee'
Error handling request from 127.0.0.1
